### ISSUE TITLE
Doesn't seem to be functioning

### ISSUE THREAD

**Issue by TopBanana123** [luckman212/wgfix#2]
Is this the correct usage? I had my main WAN drop today and then come back a few moments later, my WG Tunnels stayed up on the secondary WAN

![image](https://user-images.githubusercontent.com/43852644/197892615-c17dfcd3-99d5-472e-a6c4-3a1f3908d1d8.png)

**Comment by luckman212** [comment]
@TopBanana123 Yes that looks correct. Did you make the script executable? (Step 2 from the README; `chmod +x /etc/wgfix.sh`) ?

If you did that, please go to System > Logs and filter if for the `wgfix` Process. If you can send those logs / paste them here that would be helpful.

![CleanShot 2022-10-25 at 18 43 04](https://user-images.githubusercontent.com/1992842/197896112-e0b159ea-d919-4f99-a77a-19a62a9ff89b.png)

**Comment by TopBanana123** [comment]
Thanks for replying, yeah I did make it executable. It looks like the script does function, but it never functioned during the WAN switch, and only came alive when I logged in manually and restarted the Wireguard service to try get my tunnels back on the primary WAN

16:37 is when I had the brief outage. Ignore xxx's and 111's, tried to sanitize it as good as I could

https://pastebin.com/Ds9DW5rc

**Comment by TopBanana123** [comment]
Here is the full logs from that event

https://pastebin.com/ikrQE33m

At the end at 17:07:44 is when I logged in and started messing with things

**Comment by luckman212** [comment]
@TopBanana123 I read through the whole log you put up on pastebin, but I don't see a single "WAN UP" event logged anywhere. The script only triggers a fail-back when dpinger marks a gateway as being back online. So we need to figure out why that isn't happening first.

I see a ton of weird errors in there, rulesets failing to load, interface binding errors etc, so it's a bit hard to piece together the sequence of events. Can you describe it? Was this an actual outage or were you manually testing by unplugging cables etc?

**Comment by TopBanana123** [comment]
Some kind of issue with the AT&T Gateway, the interface went to disconnected and then reconnected for some reason. The rules failing are floating rules for fq_codel QoS

The only "WAN UP" I ever see is  from Wireguard itself. Am I perhaps misunderstanding the use of the script? I have dual WAN, and my pain point is that my Wireguard tunnels don't switch back to the primary WAN after a failover, that's what this is for, right? 

I've never seen anything to the effect of "WAN UP", just stuff like the following


Oct 25 16:37:48	php-fpm	71936	/rc.filter_configure_sync: Gateway, switch to: WAN_DHCP
Oct 25 16:37:48	php-fpm	71936	/rc.filter_configure_sync: Default gateway setting AT&T Fiber ipv4 as default.

**Comment by luckman212** [comment]
Yes, that's exactly the purpose of the script.

When dpinger marks a gateway as Online (green) it triggers the `/etc/rc.gateway_alarm` script with args to indicate the gateway in question as well as whether it's a "down" (1) or "up" (0) event. You would see an indication of this in the logs, something like
```
wgfix: WAN UP: GWNAME
```

**Comment by TopBanana123** [comment]
Interesting, I don't see that at all for either of my connections, I only see it for the Wireguard interfaces, for example

WAN UP: VPS_WG_GW

My WAN connections are in a gateway group, with one set as tier 1 and the other tier 2, any issues with doing it that way?

**Comment by luckman212** [comment]
@TopBanana123 That's odd. No, nothing wrong with having the gateways in a group (that's how I have mine as well).

Do you have either of the **Disable Gateway Monitoring/Action** checkboxes enabled on the WAN1 gateway?

**Comment by TopBanana123** [comment]
I do not, its not looking for the name "WAN" is it? Because my WAN is actually called "WAN_DHCP"

Here is the general config

![image](https://user-images.githubusercontent.com/43852644/198047662-dc2f1977-3b69-4e24-8053-293f9b49a420.png)

**Comment by luckman212** [comment]
No, the script does not depend at all on the gateway or interface names. This is very puzzling. I looked again at your logs and there is just no sign of any "WAN UP" events from `rc.gateway_alarm`:

![c7882dbbdd66c17eb99840cdcb5481d058001308](https://user-images.githubusercontent.com/1992842/198051502-7de42e22-889d-44fa-aeef-e7004c54e435.png)

What version of pfSense are you running? Can you also send the output of the following command (run from ssh or via Diagnostics → Command Prompt)
```
pgrep -lf dpinger
```

**Comment by TopBanana123** [comment]
I'm on 2.6.0, just upgraded from 2.5.2 but sadly didn't test this on 2.5.2

Will the secondary WAN also cause a WAN UP to trigger? I can test that again later today by pulling a cable or rebooting the gateway

Here is that output

```[2.6.0-RELEASE][admin@fw01.home.me.com]/root: pgrep -lf dpinger
62896 /usr/local/bin/dpinger -S -r 0 -i Mullvad_WG_Gateway -B 10.111.1111.68 -p /var/run/dpinger_Mullvad_WG_Gateway~10.111.1111.68~96.44.191.132.pid -u /var/run/dpinger_Mullvad_WG_Gateway~10.111.1111.68~96.44.191.132.sock -C /etc/rc.gateway_alarm -d 1 -s 500 -l 2000 -t 10000 -A 1000 -D 1000 -L 70 96.44.191.132
62450 /usr/local/bin/dpinger -S -r 0 -i VPS_WG_GW -B 10.9.0.4 -p /var/run/dpinger_VPS_WG_GW~10.9.0.4~10.9.0.1.pid -u /var/run/dpinger_VPS_WG_GW~10.9.0.4~10.9.0.1.sock -C /etc/rc.gateway_alarm -d 1 -s 2000 -l 2000 -t 10000 -A 2500 -D 600 -L 60 10.9.0.1
62202 /usr/local/bin/dpinger -S -r 0 -i WAN_5G_DHCP -B 75.111.1111.228 -p /var/run/dpinger_WAN_5G_DHCP~75.111.1111.228~66.174.46.49.pid -u /var/run/dpinger_WAN_5G_DHCP~75.111.1111.228~66.174.46.49.sock -C /etc/rc.gateway_alarm -d 1 -s 1500 -l 3000 -t 60000 -A 1500 -D 500 -L 60 66.174.46.49
61678 /usr/local/bin/dpinger -S -r 0 -i WAN_DHCP6 -B 2600:ippppvvv66666:ae1f:6bff:fe7b:2b5a -p /var/run/dpinger_WAN_DHCP6~2cfa2aee~2001:ippppvvv66666:69:235:123:208.pid -u /var/run/dpinger_WAN_DHCP6~2cfa2aee~2001:ippppvvv66666:69:235:123:208.sock -C /etc/rc.gateway_alarm -d 1 -s 500 -l 1500 -t 10000 -A 1000 -D 500 -L 20 2001:ippppvvv66666:69:235:123:208
61533 /usr/local/bin/dpinger -S -r 0 -i WAN_DHCP -B 104.111.111.82 -p /var/run/dpinger_WAN_DHCP~104.111.111.82~162.197.192.1.pid -u /var/run/dpinger_WAN_DHCP~104.111.111.82~162.197.192.1.sock -C /etc/rc.gateway_alarm -d 1 -s 500 -l 1500 -t 10000 -A 1000 -D 500 -L 60 162.197.192.1
```

**Comment by luckman212** [comment]
Okay, it's possible that there's an issue in 2.6.0 related to multiwan that I'm not seeing (I'm running pfS+ 22.05). But I don't think so.

Can you edit your comment above and put the log in between code fences (```)— the wrapping makes it hard to read.

**Comment by TopBanana123** [comment]
Wow, wish I'd known about that wrapping before! Done

**Comment by luckman212** [comment]
I think this might be an XY problem.

Off the top of my head, your dpinger parameters seem a little unusual to me. Why do you have such high loss percentages (60-70%) specified? And the time window of `10000` (10s) is pretty short vs the default of 60s.

I would back up a step and try putting some of these back to their defaults, then test slowly, adjusting 1 parameter at a time to make sure your gateway switching & monitoring is working in general before introducing `wgfix` back into the mix.

**Comment by TopBanana123** [comment]
The reason they have high loss percentages and shorter times is that the defaults would sometimes cause premature failover to secondary WAN. There is this weird issue with AT&T Fiber that myself or no one has been able to solve, where downloading from Steam causes packet loss to some AT&T IP's (Including the monitor IP), but no real packloss to the internet at large. So I used to be downloading a game, pfsense would flip out and switch to the 5G Backed WAN, then the packet loss would go to 0, and it would flip back, and repeat all over again

Would that impact wgfix though?

**Comment by TopBanana123** [comment]
I've switched the majority back to default, some don't need to be quite so high anymore, I'll see if that changes anything

**Comment by luckman212** [comment]
> _Would that impact wgfix though?_

Well, there's a lock mechanism implemented in wgfix to prevent it from spawning while another instance is already running. If your gateways are flapping up and down too quickly, it's possible that the script would not be able to process a changeover because the previous one was still running. But, you should _still_ see `wgfix.sh called, args: WAN_DHCP 0` entries in your logs, because that happens _before_ the lock is checked.

In general, any time your gateways are flapping that rapidly, you will have issues with pfSense; the gateway switching just doesn't work very well in those scenarios. So you should try to address the real source of the problem (AT&T).

Have you tried using a different monitor IP? I can't even ping 162.197.192.1 by the way.

**Comment by TopBanana123** [comment]
The reasons the thresholds were higher is to prevent the flapping, then numbers I had in there would just report high packet loss and keep the gateway up since it was still 100% functional. I'd love to get AT&T to fix the issue, but actually getting someone who knows networking to look at the issue is tough, if not impossible. 162.197.192.1 is about 2 hops from my local gateway, I can ping it just fine. It was identified as a good candidate by hopfinder.sh and seems pretty good. Any closer and it stays up when there has been an AT&T outage, and any further and the latency numbers are out of whack and incorrect. I believe its some local AT&T Infrastructure here in the area

**Comment by luckman212** [comment]
Ok fair enough, one more thing you can try (mentioned in the README) is manually triggering the script after a failback, just to see if it is generally working at all. If that does get your tunnel back to WAN1, then we can proceed to try and figure out what's going off the rails with dpinger/rc.gateway_alarm.

```sh
/etc/wgfix.sh WAN_DHCP 0
```

**Comment by TopBanana123** [comment]
Thanks, I actually tried the manual test before and had no luck, I figured I just was doing something wrong

If I do /etc/wgfix.sh WAN_5G_DHCP 0 would it move the tunnels to the secondary? Would be much easier to test

**Comment by luckman212** [comment]
No. As far as I know the only way to push the tunnels to the secondary (short of pulling the cable) would be to change the system's default gateway in System → Routing or by adding a static route, and then bouncing the WG service. WireGuard uses the system's routing table to determine how to route connections.

Try forcing your tunnel to the 5G and then, when you're sure your WAN1 gateway is fully online (0% packet loss), execute `/etc/wgfix.sh WAN_DHCP 0` and check your logs.

**Comment by TopBanana123** [comment]
Luckily for us, AT&T just did some maintenance and dropped my connection!

Now, both tunnels are on the 5G WAN, note the latency. Latency over the main WAN for these tunnels is around 7ms

![image](https://user-images.githubusercontent.com/43852644/198153523-5985dbb6-8d0c-48fe-875c-b656c78c3f14.png)

I did /etc/wgfix.sh WAN_DHCP 0, and nothing changed

```[2.6.0-RELEASE][admin@fw01.home.xxxxxxxx.com]/root: /etc/wgfix.sh WAN_DHCP 0
/etc/wgfix.sh called, args: WAN_DHCP 0
acquiring lockfile /tmp/wgfix.sh_WAN_DHCP.lock
WAN UP: WAN_DHCP
pausing 10s to allow gateway change to occur
checking mullvad@ (tun_wg0) at us235-wireguard.mullvad.net:51820
Route to mullvad@ endpoint (us235-wireguard.mullvad.net) is via interface: igb2 (104.111.111.82)
Looking for states to us235-wireguard.mullvad.net:51820 that are NOT related to 104.111.111.82
no bad states found
checking linode@ (tun_wg1) at wg-vps1.xxxxxxxx.com:51822
Route to linode@ endpoint (wg-vps1.xxxxxxxx.com) is via interface: igb2 (104.111.111.82)
Looking for states to wg-vps1.xxxxxxxx.com:51822 that are NOT related to 104.111.111.82
no bad states found
done, removing lockfile /tmp/wgfix.sh_WAN_DHCP.lock
[2.6.0-RELEASE][admin@fw01.home.xxxxxxxx.com]/root:
```

**Comment by TopBanana123** [comment]
Confirmation its going over the secondary WAN

![image](https://user-images.githubusercontent.com/43852644/198155962-a4043170-6776-44d4-89ce-dab35c26424d.png)

**Comment by luckman212** [comment]
@TopBanana123 That last bit of logging was helpful. I think I know what the problem is here, I'll try to come up with a fix shortly.

**Comment by TopBanana123** [comment]
Awesome, thanks! I really appreciate it

**Comment by jstride** [comment]
Did this need an update in the end? If I execute /etc/wgfix.sh WAN_DHCP 0, I get the following, should it be finding states as traffic is still following over the 4G backup interface?:

`/etc/wgfix.sh called, args: WAN_DHCP 0
acquiring lockfile /tmp/wgfix.sh_WAN_DHCP.lock
WAN UP: WAN_DHCP
pausing 10s to allow gateway change to occur
checking PER@ (tun_wg0) at xxx:51820
Route to PER endpoint (xxx) is via interface: lagg0.4090 (xxx)
Looking for states to xxx:51820 that are NOT related to xxx
no bad states found
checking SYD@ (tun_wg1) at xxx:51820
Route to SYD endpoint (xxx) is via interface: lagg0.4090 (xxx)
Looking for states to xxx:51820 that are NOT related to xxx
no bad states found
done, removing lockfile /tmp/wgfix.sh_WAN_DHCP.lock`

I don't seem to see any WAN UP events in the logs either. I am running 23.05.1-RELEASE

**Comment by luckman212** [comment]
23.09 is coming in a few days, I'm going to hold off on any work here until that releases and then I plan to re-test and update as needed. I just re-read this entire thread and I wish I had posted back a year ago when I seemed to be aware of what needed to be fixed! But now I'll have to refresh my memory. I'll flag this for a look after 23.09 drops.
